File: _inc/laravel/app/Http/Controllers/Companies/CompanyPolicyController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    CompaniesConstants,
    DatabaseConstants,
    UsersConstants,
    ViewsConstants,
};
use App\Models\{Branch, CompanyPolicy, Utility};
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Http\{
    JsonResponse,
    RedirectResponse,
    Request
};
use Illuminate\Support\Facades\{
    File,
    Log,
    Route,
    Validator,
    View as ViewFacade,
};

final class CompanyPolicyController extends Controller
{
    use ChecksLogin, ChecksPermissions;

    private const REDIRECT_INDEX = '/';

    public function index(Request $r)
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($r, $action, $method, $class, $base) {
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($c = self::guard($r, 'manage company policy', self::REDIRECT_INDEX)) !== true) return $c; // ! ALERT
            Log::debug("[{$base}::{$action}] start", [UsersConstants::COL_USER_ID => $r->user()?->id, 'method' => $method]);
            try {
                $qStart = microtime(true);
                $companyPolicy = CompanyPolicy::where(DatabaseConstants::TABLE_CREATOR, $u->creatorId())->with('branches')->get();
                $this->logExecutionTime($qStart, $action, 'fetchPolicies');
                $viewPath = ViewsConstants::CPN_PL . '.' . $action;
                if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                $renderStart = microtime(true);
                $resp = view($viewPath, compact('companyPolicy'));
                $this->logExecutionTime($renderStart, $action, 'renderView');
                return $resp;
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($r, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function create(Request $r)
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($r, $action, $method, $class, $base) {
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($c = self::guard($r, 'create company policy', self::REDIRECT_INDEX)) !== true) return $c; // ! ALERT
            Log::debug("[{$base}::{$action}] start", [UsersConstants::COL_USER_ID => $r->user()?->id, 'method' => $method]);
            $branch = self::branches($u->creatorId());
            $viewPath = ViewsConstants::CPN_PL . '.' . $action;
            if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            $renderStart = microtime(true);
            $resp = view($viewPath, compact('branch'));
            $this->logExecutionTime($renderStart, $action, 'renderView');
            return $resp;
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function store(Request $r): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($r, $action, $method, $class, $base) {
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($c = self::guard($r, 'create company policy', self::REDIRECT_INDEX)) !== true) return $c; // ! ALERT
            if ($c = self::v($r, ['branch' => 'required', 'title' => 'required'])) return $c;
            try {
                $uplStart = microtime(true);
                $file = $r->hasFile('attachment') ? self::upload($r, 'attachment') : null;
                $this->logExecutionTime($uplStart, $action, 'uploadAttachment');
                $crtStart = microtime(true);
                $policy = CompanyPolicy::create([
                    'branch' => $r->branch,
                    'title' => $r->title,
                    'description' => $r->description,
                    'attachment' => $file,
                    DatabaseConstants::TABLE_CREATOR => $u->creatorId()
                ]);
                $this->logExecutionTime($crtStart, $action, 'createPolicy');
                try {
                    $settings = Utility::settings($u->creatorId());
                    $branch = Branch::find($r->branch);
                    $payload = [
                        'company_policy_name' => $policy->title,
                        'branch_name' => $branch?->name ?? ''
                    ];
                    ($settings['policy_notification'] ?? false) && Utility::sendSlackMsg('new_company_policy', $payload);
                    ($settings['telegram_policy_notification'] ?? false) && Utility::sendTelegramMsg('new_company_policy', $payload);
                } catch (\Throwable $e) {
                    Log::warning($class . '::' . $action . ' notify ' . $e->getMessage());
                }
                try {
                    if ($hook = Utility::webhookSetting('New Company Policy')) {
                        $ok = Utility::webhookCall($hook['url'], $policy->toJson(), $hook['method']);
                        $ok || Log::warning($class . '::' . $action . ' webhook failed');
                    }
                } catch (\Throwable $e) {
                    Log::warning($class . '::' . $action . ' webhook ' . $e->getMessage());
                }
                Log::info("[{$base}::{$action}] created", ['policy_id' => $policy->id, UsersConstants::COL_USER_ID => $r->user()?->id]);
                return redirect()->route(ViewsConstants::CPN_PL . '.index')->with('success', __('Company policy successfully created.'));
            } catch (\RuntimeException $e) {
                return back()->with('error', $e->getMessage());
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($r, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function edit(Request $r, CompanyPolicy $companyPolicy)
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($r, $companyPolicy, $action, $method, $class, $base) {
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($c = self::guard($r, 'edit company policy', self::REDIRECT_INDEX)) !== true) return $c; // ! ALERT
            $branch = self::branches($u->creatorId());
            $viewPath = ViewsConstants::CPN_PL . '.' . $action;
            if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            $renderStart = microtime(true);
            $resp = view($viewPath, compact('branch', 'companyPolicy'));
            $this->logExecutionTime($renderStart, $action, 'renderView');
            return $resp;
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'policy_id' => $companyPolicy->id]);
    }

    public function update(Request $r, CompanyPolicy $companyPolicy): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($r, $companyPolicy, $action, $method, $class, $base) {
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($c = self::guard($r, 'edit company policy', self::REDIRECT_INDEX)) !== true) return $c; // ! ALERT
            if ($c = self::v($r, ['branch' => 'required', 'title' => 'required'])) return $c;
            try {
                $data = [
                    'branch' => $r->branch,
                    'title' => $r->title,
                    'description' => $r->description
                ];
                if ($r->hasFile('attachment')) {
                    $uplStart = microtime(true);
                    $data['attachment'] = self::upload($r, 'attachment');
                    $this->logExecutionTime($uplStart, $action, 'uploadAttachment');
                }
                $updStart = microtime(true);
                $companyPolicy->update($data);
                $this->logExecutionTime($updStart, $action, 'updatePolicy');
                Log::info("[{$base}::{$action}] updated", ['policy_id' => $companyPolicy->id, UsersConstants::COL_USER_ID => $r->user()?->id]);
                return redirect()->route(ViewsConstants::CPN_PL . '.index')->with('success', __('Company policy successfully updated.'));
            } catch (\RuntimeException $e) {
                return back()->with('error', $e->getMessage());
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage(), 'policy_id' => $companyPolicy->id]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($r, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'policy_id' => $companyPolicy->id]);
    }

    public function destroy(Request $r, CompanyPolicy $companyPolicy): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($r, $companyPolicy, $action, $method, $class, $base) {
            if (($u = self::_checkLogin()) instanceof RedirectResponse) return $u;
            if (($c = self::guard($r, 'delete document', self::REDIRECT_INDEX)) !== true) return $c; // ! ALERT
            if ($companyPolicy->created_by !== $u->creatorId()) return back()->with('error', __('Permission denied.'));
            try {
                if ($companyPolicy->attachment) {
                    $file = 'uploads/companyPolicy/' . $companyPolicy->attachment;
                    File::exists($file) && File::delete($file);
                }
                $delStart = microtime(true);
                $companyPolicy->delete();
                $this->logExecutionTime($delStart, $action, 'deletePolicy');
                Log::info("[{$base}::{$action}] deleted", ['policy_id' => $companyPolicy->id, UsersConstants::COL_USER_ID => $r->user()?->id]);
                return redirect()->route(ViewsConstants::CPN_PL . '.index')->with('success', __('Company policy successfully deleted.'));
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage(), 'policy_id' => $companyPolicy->id]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($r, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'policy_id' => $companyPolicy->id]);
    }

    public function show(): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () {
            return redirect()->route(ViewsConstants::CPN_PL . '.index');
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    private static function v(Request $r, array $rules): ?RedirectResponse
    {
        $v = Validator::make($r->all(), $rules);
        return $v->fails()
            ? redirect()->back()->with('error', $v->getMessageBag()->first())
            : null;
    }

    private static function branches(int $creator): array
    {
        try {
            return Branch::where(DatabaseConstants::TABLE_CREATOR, $creator)
                ->pluck(CompaniesConstants::COL_BRC_NM, 'id')
                ->prepend(__('Select Branch'), '')
                ->all();
        } catch (\Throwable $e) {
            Log::error(__CLASS__ . '::' . __FUNCTION__ . ' ' . $e->getMessage());
            return [];
        }
    }

    private static function upload(
        Request $r,
        string  $field,
        string  $dir = 'uploads/companyPolicy/'
    ): string {
        $file     = $r->file($field);
        $nameParts = pathinfo($file->getClientOriginalName());
        $name     = "{$nameParts['filename']}_" . time() . '.' . $nameParts['extension'];
        try {
            $path = Utility::uploadFile($r, $field, $name, $dir, []);
            return $path['flag'] ? $name : throw new \RuntimeException($path['msg']);
        } catch (\Throwable $e) {
            throw new \RuntimeException('upload', 0, $e); // ! ALERT bubbling msg
        }
    }
}

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/companies/company_policy_controller.py
import json
import os
import time
from django.contrib import messages
from django.http import HttpRequest, HttpResponse, JsonResponse
from django.shortcuts import get_object_or_404, redirect, render
from ....Models.companies.branch import Branch
from ....Models.companies.company_policy import CompanyPolicy
from ....Models.utils.utility import Utility
from .._traits.controller import Controller
from .._helpers.http import get_redirect_url
from django.core.exceptions import PermissionDenied
from ....configs.messages_templates import get_exception_class_message

class CompanyPolicyController(Controller):

  def index(self, request: HttpRequest) -> HttpResponse:
    try:
      return (render(request, 'companyPolicy/index.html', {'companyPolicy': CompanyPolicy.objects.filter(created_by=request.user.creator_id()).prefetch_related('branches')})
              if request.user.has_perm('manage company policy')
              else (messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__)) or redirect(get_redirect_url(request))))
    except Exception as e:
      print(f"[CompanyPolicyController.index] Failed: {e.__class__.__name__}: {e}")
      messages.error(request, "CompanyPolicyController.index: Error")
      return redirect(get_redirect_url(request))

  def create(self, request: HttpRequest) -> HttpResponse:
    try:
      return (render(request, 'companyPolicy/create.html', {'branch': {'': 'Select Branch', **{b.id: b.name for b in Branch.objects.filter(created_by=request.user.creator_id())}}})
              if request.user.has_perm('create company policy')
              else JsonResponse({'error': "Permission denied."}, status=401))
    except Exception as e:
      print(f"[CompanyPolicyController.create] Failed: {e.__class__.__name__}: {e}")
      messages.error(request, "CompanyPolicyController.create: Error")
      return redirect(get_redirect_url(request))

  def store(self, request: HttpRequest) -> HttpResponse:
    try:
      if not request.user.has_perm('create company policy'):
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
        return redirect(get_redirect_url(request))
      branch_val = request.POST.get('branch')
      title = request.POST.get('title')
      if not branch_val or not title:
        messages.error(request, "Branch and Title are required.")
        return redirect(get_redirect_url(request))
      fileNameToStore = ''
      if request.FILES.get('attachment'):
        file_obj = request.FILES['attachment']
        originalName = file_obj.name
        filename = os.path.splitext(originalName)[0]
        extension = os.path.splitext(originalName)[1].lstrip('.')
        fileNameToStore = f"{filename}_{int(time.time())}.{extension}"
        dir_path = 'uploads/companyPolicy/'
        image_path = os.path.join(dir_path, fileNameToStore)
        if os.path.exists(image_path):
          os.remove(image_path)
        upload_result = Utility.upload_file(request, 'attachment', fileNameToStore, dir_path, [])
        if upload_result.get('flag') != 1:
          messages.error(request, upload_result.get('msg'))
          return redirect(get_redirect_url(request))
      policy = CompanyPolicy()
      policy.branch = branch_val
      policy.title = title
      policy.description = request.POST.get('description')
      policy.attachment = fileNameToStore if request.FILES.get('attachment') else ''
      policy.created_by = request.user.creator_id()
      policy.save()
      setting = Utility.settings(request.user.creator_id())
      branch_obj = get_object_or_404(Branch, pk=branch_val)
      policyNotificationArr = {'company_policy_name': title, 'branch_name': branch_obj.name}
      if setting.get('policy_notification') == 1:
        Utility.send_slack_msg('new_company_policy', policyNotificationArr)
      if setting.get('telegram_policy_notification') == 1:
        Utility.send_telegram_msg('new_company_policy', policyNotificationArr)
      webhook = Utility.webhookSetting('New Company Policy')
      if webhook:
        parameter = json.dumps(policy.__dict__)
        status = Utility.WebhookCall(webhook.get('url'), parameter, webhook.get('method'))
        if status:
          messages.success(request, "Company policy successfully created.")
          return redirect('company-policy_index')
        else:
          messages.error(request, "Webhook call failed.")
          return redirect(get_redirect_url(request))
      messages.success(request, "Company policy successfully created.")
      return redirect('company-policy_index')
    except Exception as e:
      print(f"[CompanyPolicyController.store] Failed: {e.__class__.__name__}: {e}")
      messages.error(request, "CompanyPolicyController.store: Error")
      return redirect(get_redirect_url(request))

  def show(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      return render(request, 'companyPolicy/show.html', {})
    except Exception as e:
      print(f"[CompanyPolicyController.show] Failed: {e.__class__.__name__}: {e}")
      messages.error(request, "CompanyPolicyController.show: Error")
      return redirect(get_redirect_url(request))

  def edit(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      return (render(request, 'companyPolicy/edit.html', {'branch': {'': 'Select Branch', **{b.id: b.name for b in Branch.objects.filter(created_by=request.user.creator_id())}}, 'companyPolicy': get_object_or_404(CompanyPolicy, pk=id)})
              if request.user.has_perm('edit company policy')
              else JsonResponse({'error': "Permission denied."}, status=401))
    except Exception as e:
      print(f"[CompanyPolicyController.edit] Failed: {e.__class__.__name__}: {e}")
      messages.error(request, "CompanyPolicyController.edit: Error")
      return redirect(get_redirect_url(request))

  def update(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      if not request.user.has_perm('create company policy'):
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
        return redirect(get_redirect_url(request))
      branch_val = request.POST.get('branch')
      title = request.POST.get('title')
      if not branch_val or not title:
        messages.error(request, "Branch and Title are required.")
        return redirect(get_redirect_url(request))
      fileNameToStore = ''
      if request.FILES.get('attachment'):
        file_obj = request.FILES['attachment']
        originalName = file_obj.name
        filename = os.path.splitext(originalName)[0]
        extension = os.path.splitext(originalName)[1].lstrip('.')
        fileNameToStore = f"{filename}_{int(time.time())}.{extension}"
        dir_path = 'uploads/companyPolicy/'
        image_path = os.path.join(dir_path, fileNameToStore)
        if os.path.exists(image_path):
          os.remove(image_path)
        upload_result = Utility.upload_file(request, 'attachment', fileNameToStore, dir_path, [])
        if upload_result.get('flag') != 1:
          messages.error(request, upload_result.get('msg'))
          return redirect(get_redirect_url(request))
      companyPolicy = get_object_or_404(CompanyPolicy, pk=id)
      companyPolicy.branch = branch_val
      companyPolicy.title = title
      companyPolicy.description = request.POST.get('description')
      if request.FILES.get('attachment'):
        companyPolicy.attachment = fileNameToStore
      companyPolicy.created_by = request.user.creator_id()
      companyPolicy.save()
      messages.success(request, "Company policy successfully updated.")
      return redirect('company-policy_index')
    except Exception as e:
      print(f"[CompanyPolicyController.update] Failed: {e.__class__.__name__}: {e}")
      messages.error(request, "CompanyPolicyController.update: Error")
      return redirect(get_redirect_url(request))

  def destroy(self, request: HttpRequest, id: int) -> HttpResponse:
    try:
      if not request.user.has_perm('delete document'):
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
        return redirect(get_redirect_url(request))
      companyPolicy = get_object_or_404(CompanyPolicy, pk=id)
      if companyPolicy.created_by != request.user.creator_id():
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))
        return redirect(get_redirect_url(request))
      companyPolicy.delete()
      dir_path = os.path.join('uploads/companyPolicy/')
      if companyPolicy.attachment:
        file_path = os.path.join(dir_path, companyPolicy.attachment)
        if os.path.exists(file_path):
          os.unlink(file_path)
      messages.success(request, "Company policy successfully deleted.")
      return redirect('company-policy_index')
    except Exception as e:
      print(f"[CompanyPolicyController.destroy] Failed: {e.__class__.__name__}: {e}")
      messages.error(request, "CompanyPolicyController.destroy: Error")
      return redirect(get_redirect_url(request))
